src/HappyEyeBallsConnectionBuilder.php

```diff
@@ -0,0 +1,311 @@
+<?php
+
+namespace React\Socket;
+
+use React\Dns\Model\Message;
+use React\Dns\Resolver\ResolverInterface;
+use React\EventLoop\LoopInterface;
+use React\EventLoop\TimerInterface;
+use React\Promise;
+use React\Promise\CancellablePromiseInterface;
+
+/**
+ * @internal
+ */
+final class HappyEyeBallsConnectionBuilder
+{
+    const CHECK_INTERVAL = 0.1;
+
+    public $loop;
+    public $connector;
+    public $resolver;
+    public $uri;
+    public $host;
+    public $resolved = array(
+        Message::TYPE_A    => false,
+        Message::TYPE_AAAA => false,
+    );
+    public $resolverPromises = array();
+    public $connectionPromises = array();
+    public $connectQueue = array();
+    public $timer;
+    public $parts;
+    public $ipsCount = 0;
+    public $failureCount = 0;
+    public $resolve;
+    public $reject;
+
+    public function __construct(LoopInterface $loop, ConnectorInterface $connector, ResolverInterface $resolver, $uri, $host, $parts)
+    {
+        $this->loop = $loop;
+        $this->connector = $connector;
+        $this->resolver = $resolver;
+        $this->uri = $uri;
+        $this->host = $host;
+        $this->parts = $parts;
+    }
+
+    public function connect()
+    {
+        $that = $this;
+        return new Promise\Promise(function ($resolve, $reject) use ($that) {
+            $lookupResolve = function ($type) use ($that, $resolve, $reject) {
+                return function (array $ips) use ($that, $type, $resolve, $reject) {
+                    unset($that->resolverPromises[$type]);
+                    $that->resolved[$type] = true;
+
+                    $that->mixIpsIntoConnectQueue($ips);
+
+                    if ($that->timer instanceof TimerInterface) {
+                        return;
+                    }
```

Comment:
This check causes the bug discussed above. The timer will only be started when more IPs are available when the first connection attempt (IPv6) is started. If the IPv4 address resolves afterwards, then this timer is not running and it will immediately try the IPv4 connection instead of waiting first.

Reply:
Can you take a look at this again? I can't reproduce this locally anymore, but I don't see any change here?

Reply:
This was solved in the check method.